Improve documentation for v5

davidtheclark opened the pull request:
I took a pass through the README documentation.

I also added a remark-cli step that lints the README (with autocorrect when possible). I've found this especially nice to catch broken links.

@sudo-suhas or @olsonpm for review, please.

olsonpm commented on README.md:
```diff
@@ -29,99 +31,99 @@ Tested in Node 4+.
 If `false`, cosmiconfig will not look for an rc file.
 
-If `rcExtensions: true`, the rc file can also have extensions that specify the syntax, e.g. `.[moduleName]rc.json`.
+If `rcExtensions: true`, the rc file can also have extensions that specify the syntax, e.g. `.${moduleName]rc.json`.
```
```txt
${moduleName]rc.json
            ^
```

olsonpm commented on README.md:
```diff
@@ -145,104 +147,129 @@ Instead of *just* looking for `.goldengrahamsrc` (no extension), it will also lo
-Loads a configuration file directly.  Rejects with an error if the file does not exist, if the file contents are empty, or if the file cannot be parsed.  Upon success it returns a Promise that resolves with an object with two properties:
-- `config`: The loaded and parsed configuration.
-- `filepath`: The filepath where this configuration was found.
+By default, `search`, ignores empty configuration files and continues searching up the tree. If the `search` option [`ignoreEmpty`] is set to `false` and an empty configuration file is found, the [result] will have include `config: undefined` and `isEmpty: true`.
```
I'm the worst with commas, but just checking to see if you intended on the comma after `search`

```txt
By default, `search`, ignores
                    ^
```

olsonpm commented on README.md:
````diff
@@ -29,99 +31,99 @@ Tested in Node 4+.
+```
 
-Creates a cosmiconfig instance (i.e. explorer) configured according to the arguments, and initializes its caches.
+Creates a cosmiconfig instance ("explorer") configured according to the arguments, and initializes its caches.
````
Not sure whether you and @sudosuhas already discussed this, but do you have thoughts on normalizing the terminology surrounding **cosmiconfig instance** and **explorer**?  Another term used right under '## Usage' is **cosmiconfig explorer**.

This might be solved by an api change with the method 'createExplorer' being a property of cosmiconfig.  Anything besides that leaves room for ambiguity - which is not a big deal.  Just wanted to see what you guys thought.

davidtheclark replied:
I don't think we need to make any API change; but if we wanted to change the docs we could do this:

```js
// Change this:
const cosmiconfig = require('cosmiconfig');
// To this:
const createExplorer = require('cosmiconfig');
```

olsonpm replied:
haha true

sudo-suhas commented on README.md:
```diff
@@ -29,99 +31,99 @@ Tested in Node 4+.
+    // result.filepath is the path to the config file that was found.
+    // result.isEmpty is true if there was nothing to parse in the config file.
+  },
+  (parsingError) => {
```
I prefer the `.catch` equivalent as it seems clearer and explicit(to me at least).

davidtheclark replied:
It's less precise, though: if your callback that does something with `result` throws an error, that error is going to end up in the `catch` statement but will not be a parsing error.

davidtheclark replied:
I'll change it to `catch` and just change the `parsingError` name to `error`.

sudo-suhas commented on README.md:
```diff
@@ -29,99 +31,99 @@ Tested in Node 4+.
-Type: `string` or `false`
-Default: `'[moduleName]'`
+Type: `string` or `false`.
+Default: `'${moduleName}'`.
```
This still renders to a single line. If we want the `Type` and `Default` to be on different lines, we should either use `<br>` or bulletting.

davidtheclark replied:
Yeah, I think that's fine. I often type Markdown with individual sentences on separate lines to make the diff more meaningful.

sudo-suhas commented on README.md:
```diff
@@ -29,99 +31,99 @@ Tested in Node 4+.
-Type: `string` or `false`
-Default: `'.[moduleName]rc'`
+Type: `string` or `false`.
+Default: `\`.${moduleName}rc\`\`.
```
The escaping does not work. This is how it renders:

![image](https://user-images.githubusercontent.com/22251956/37264799-8ba937ea-25d5-11e8-8878-2d4bc02bdeb0.png)

davidtheclark replied:
Bummer.

sudo-suhas commented on README.md:
```diff
@@ -29,99 +31,99 @@ Tested in Node 4+.
-Type: `string` or `false`
-Default: `'[moduleName].config.js'`
+Type: `string` or `false`.
+Default: `\`${moduleName}.config.js\`\`.
```
Escaping needs to be removed here too.

sudo-suhas commented on README.md:
```diff
@@ -145,104 +147,129 @@ Instead of *just* looking for `.goldengrahamsrc` (no extension), it will also lo
+Searches for a configuration file. In async mode (default), returns a Promise that resolves a [result] object or with `null`, if no configuration file is found. In [`sync`] mode, returns a [result] object or `null`.
 
-When the [search option](#searchoptions) `ignoreEmpty` is set to false and an empty configuration file is found, the result will have `config` set to `undefined` and an additional property `isEmpty` set to true.
+So let's say `const explorer = cosmiconfig('goldengrahams');` — here's how `explorer.search` will work:
```
Change to "here's how `explorer.search` works" to be consistent with the rest of the explanation.

sudo-suhas commented on README.md:
```diff
@@ -145,104 +147,129 @@ Instead of *just* looking for `.goldengrahamsrc` (no extension), it will also lo
-Loads a configuration file directly.  Rejects with an error if the file does not exist, if the file contents are empty, or if the file cannot be parsed.  Upon success it returns a Promise that resolves with an object with two properties:
-- `config`: The loaded and parsed configuration.
-- `filepath`: The filepath where this configuration was found.
+By default, `search` ignores empty configuration files and continues searching up the tree. If the `search` option [`ignoreEmpty`] is set to `false` and an empty configuration file is found, the [result] will have include `config: undefined` and `isEmpty: true`.
```
`"will have include" => "will have"`

sudo-suhas commented on README.md:
```diff
@@ -145,104 +147,129 @@ Instead of *just* looking for `.goldengrahamsrc` (no extension), it will also lo
+
+## Caching
+
+As of v2, cosmiconfig uses a few caches to reduce the need for repetitious reading of the filesystem. Every new cosmiconfig instance (created with `cosmiconfig()`) has its own caches.
```
`"uses a few caches" => "uses caching"`

sudo-suhas commented on README.md:
```diff
@@ -29,99 +31,99 @@ Tested in Node 4+.
-- use the cache clearing methods documented below.
+- **config:** The parsed configuration object. `undefined` if the file is empty.
+- **filepath:** The path to the configuration file that was found.
+- **isEmpty:** `true` if the configuration file is empty.
```
Should we mention that this property will not be present if the search was successful?

davidtheclark replied:
Sure.

sudo-suhas commented:
What do you think of adding a "Used By" section to the readme?

sudo-suhas commented on CHANGELOG.md:
```diff
@@ -26,7 +26,7 @@
 - Removed: support for loading config path using the `--config` flag. cosmiconfig will not parse command line arguments. Your application can parse command line arguments and pass them to cosmiconfig.
 - Removed: `argv` config option.
-- Removed: support for Node versions < 4.
+- Removed: support for Node versions &lt; 4.
```
Is it possible to use the `<` symbol instead of `&lt;`? Seems unnecessary..

davidtheclark replied:
This is automatically changed by remark-cli, so I'm not going to worry about it.

davidtheclark commented:
@sudo-suhas Thanks for catching those typos and for the suggestions. I addressed the feedback in the new commit.

If you'd like to add a "Used By" section to the README, sounds good to me 👍

sudo-suhas commented on README.md:
```diff
@@ -145,104 +146,129 @@ Instead of *just* looking for `.goldengrahamsrc` (no extension), it will also lo
+Default: `true`.
 
-#### `load([configPath])`
+By default, `search` ignores empty configuration files and continues searching up the tree. If the `search` option [`ignoreEmpty`] is set to `false` and an empty configuration file is found, the [result] will include `config: undefined` and `isEmpty: true`.
```
This links back to itself. Is that necessary?
